pkg/tsdb/cloudwatch/routes/namespaces_test.go

```diff
@@ -0,0 +1,83 @@
+package routes
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/tsdb/cloudwatch/models"
+	"github.com/grafana/grafana/pkg/tsdb/cloudwatch/services"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_Namespaces_Route(t *testing.T) {
+	customNamespaces := ""
+	factoryFunc := func(pluginCtx backend.PluginContext, region string) (reqCtx models.RequestContext, err error) {
+		return models.RequestContext{
+			Settings: &models.CloudWatchSettings{
+				Namespace: customNamespaces,
+			},
+		}, nil
+	}
+
+	t.Run("calls GetHardCodedNamespaces", func(t *testing.T) {
+		origGetHardCodedNamespaces := services.GetHardCodedNamespaces
+		t.Cleanup(func() {
+			services.GetHardCodedNamespaces = origGetHardCodedNamespaces
+		})
+		haveBeenCalled := false
+		services.GetHardCodedNamespaces = func() []string {
+			haveBeenCalled = true
+			return []string{}
+		}
+		rr := httptest.NewRecorder()
+		req := httptest.NewRequest("GET", "/namespaces", nil)
+		handler := http.HandlerFunc(ResourceRequestMiddleware(NamespacesHandler, factoryFunc))
+		handler.ServeHTTP(rr, req)
+		res := []models.Metric{}
+		err := json.Unmarshal(rr.Body.Bytes(), &res)
```

Comment:
Seems like the mock return value + this assertion might not be quite right/robust enough.

[]models.Metric is the incorrect type to unmarshal to, but I think there is no json.Unmarshal error because the mock return value is just the empty slice of []string{}.
You can see the error if you try the real `services.GetHardCodedNamespaces` (for example: comment the stub, run the test).

I'm thinking:
* return real values from the mock GetHardCodedNamespaces and assert those values are in the result `res`. This may make this test a little less "dumb," but I think just one hardcoded namespace, no custom namespace, could be simple enough?
* assert the JSON string (assert.JsonEq) instead of making this Unmarshal step
* won't need to signal mock hasBeenCalled because we'll be asserting the mocked return value. that'll keep the mock lightweight
* Were there some existing tests on the /namespaces endpoint? Didn't see any around metric_find_query.go

Reply:
Nice catch! I'll remove the json parsing here. As the test name suggests, this only assert that the method is being called. What happens with returned namespaces is already tested in other tests in this file